In [1]:
import importlib
import data_loader
importlib.reload(data_loader)
from data_loader import DataLoader

from constants import PATH_DATASETS, TARGET_COLUMN_CATEGORY_LEVEL2
from constants_v2 import PATH_DATASETS_V2

In [2]:
# ========================================
# STEP 1: LOAD DATA
# ========================================
print("\n" + "="*70)
print("STEP 1: LOADING DATA")
print("="*70)    
    
loader = DataLoader(
        metadata_path=f'{PATH_DATASETS}amazon_books_metadata_sample_20k.csv',
        reviews_path=f'{PATH_DATASETS}amazon_books_reviews_sample_20k.csv'
    )

# Load both datasets
metadata_df = loader.load_metadata()#nrows=1000
reviews_df = loader.load_reviews()#nrows=5000    


2025-11-09 15:38:10,003 - data_loader - INFO - DataLoader initialized
2025-11-09 15:38:10,003 - data_loader - INFO -   Metadata: datasets\amazon_books_metadata_sample_20k.csv
2025-11-09 15:38:10,004 - data_loader - INFO -   Reviews: datasets\amazon_books_reviews_sample_20k.csv
2025-11-09 15:38:10,006 - data_loader - INFO - Loading metadata from: datasets\amazon_books_metadata_sample_20k.csv



STEP 1: LOADING DATA


2025-11-09 15:38:12,492 - data_loader - INFO - Successfully loaded 20,000 rows and 28 columns in 2.48s
2025-11-09 15:38:12,494 - data_loader - INFO - Loading reviews from: datasets\amazon_books_reviews_sample_20k.csv
2025-11-09 15:38:21,262 - data_loader - INFO - Successfully loaded 727,876 reviews and 12 columns in 8.77s


In [3]:
#metadata_df.columns
# Validate both
validation = loader.validate_schema('both')

2025-11-09 15:38:21,278 - data_loader - INFO - Validating schema for: both
2025-11-09 15:38:21,285 - data_loader - INFO - ✓ Metadata schema validation passed
2025-11-09 15:38:21,287 - data_loader - INFO - Metadata optional columns: 10/10
2025-11-09 15:38:21,289 - data_loader - INFO - ✓ Reviews schema validation passed
2025-11-09 15:38:21,291 - data_loader - INFO - Reviews optional columns: 8/8


In [5]:
   # Print comprehensive summary
loader.print_summary()   

2025-11-09 15:38:27,025 - data_loader - INFO - Generating data summary...
2025-11-09 15:38:27,172 - data_loader - INFO - Data summary generated successfully
2025-11-09 15:38:27,174 - data_loader - INFO - Generating reviews summary...



📚 BOOK METADATA & REVIEWS SUMMARY

📊 METADATA - BASIC INFORMATION:
  • Total Books: 20,000
  • Total Columns: 28
  • Memory Usage: 271.43 MB
  • Duplicates: 0

🏷️  CATEGORY STATISTICS:

  category_level_2_sub:
    - Unique Categories: 50
    - Missing Values: 118
    - Top 3 Categories:
      1. Literature & Fiction: 5,174 books
      2. Mystery, Thriller & Suspense: 2,631 books
      3. Children's Books: 1,952 books

📝 TEXT FEATURES:

  title:
    - Missing: 0
    - Avg Length: 47 characters
    - Range: 1 - 200 chars

  description:
    - Missing: 0
    - Avg Length: 5154 characters
    - Range: 10 - 254372 chars

  features_text:
    - Missing: 33
    - Avg Length: 1168 characters
    - Range: 1 - 24835 chars

🔢 NUMERICAL FEATURES:

  price_numeric:
    - Count: 20,000
    - Mean: 18.78
    - Range: 5.01 - 11568.31
    - Missing: 0

  page_count:
    - Count: 15,544
    - Mean: 330.96
    - Range: 0.0 - 9998.0
    - Missing: 4,456

  average_rating:
    - Count: 20,000
    - Mean: 

2025-11-09 15:38:27,921 - data_loader - INFO - Reviews summary generated successfully



💬 REVIEWS SUMMARY

📊 BASIC STATISTICS:
  • Total Reviews: 727,876
  • Unique Users: 593,235
  • Unique Books Reviewed: 19,991
  • Average Rating: 4.42
  • Median Rating: 5.0

⭐ RATING DISTRIBUTION:
  1★ : 29,554 reviews (4.1%)
  2★ : 28,275 reviews (3.9%)
  3★ : 52,300 reviews (7.2%)
  4★ : 116,045 reviews (15.9%)
  5★ : 501,702 reviews (68.9%)

📝 REVIEW TEXT:
  • Avg Length: 432 characters
  • Median Length: 167 characters
  • Range: 1 - 32285 chars
  • Missing Text: 112

✅ VERIFIED PURCHASES:
  • Verified: 523,636 (71.94%)
  • Not Verified: 204,240

👍 HELPFUL VOTES:
  • Total Votes: 1,430,192
  • Avg per Review: 1.96
  • Reviews with Votes: 240,043

📅 TEMPORAL DISTRIBUTION:
  • Date Range: 1997 - 2023
  • Top 3 Years:
    1. 1997: 137 reviews
    2. 1998: 572 reviews
    3. 1999: 1,427 reviews




In [6]:
reviews_summary = loader.get_reviews_summary()
#print(reviews_summary)    

2025-11-09 15:38:32,417 - data_loader - INFO - Generating reviews summary...
2025-11-09 15:38:33,180 - data_loader - INFO - Reviews summary generated successfully


In [7]:

# Get reviews per book statistics
reviews_per_book = loader.get_reviews_per_book_stats()
print(reviews_per_book.head(10))

2025-11-09 15:38:35,740 - data_loader - INFO - Analyzing reviews per book...
2025-11-09 15:38:36,064 - data_loader - INFO - Analyzed 19,991 books


      parent_asin  num_reviews  avg_rating  rating_std  unique_users
19096  B00CNQ7HAU        14892    4.493956    0.807029         14752
19306  B00UEKRTW8         7799    3.986152    1.154877          7720
19183  B00G3L6KQI         4492    4.625779    0.785633          4464
19219  B00IJJUIMY         4222    4.383468    0.891159          4194
4330   0439023521         3150    4.475873    1.000741          3100
2717   0316460834         2848    4.795646    0.689391          2835
398    0061537969         2520    4.631746    0.879446          2490
19336  B0106E2CO6         2333    4.469781    0.901000          2308
8148   0811879542         2137    4.745905    0.787611          2124
13796  1565125606         2025    4.291852    1.126073          2009


In [10]:
    
# Join with minimum 5 reviews per book
joined_df = loader.join_metadata_reviews(
        how='inner',
        min_reviews_per_book=5
)
print(f'\n{"="*60}')
print(f"\nJoined data shape:\n {joined_df.shape}")
    
# Get rating distribution by category
rating_by_category = loader.get_rating_distribution_by_category(TARGET_COLUMN_CATEGORY_LEVEL2)
print(f'\n{"="*60}')
print(f"\nRating distribution by {TARGET_COLUMN_CATEGORY_LEVEL2} category:\n")
print(rating_by_category)

loader.export_summary_report(f'{PATH_DATASETS_V2}complete_data_report.txt')


print(f"\nLoaded:")
print(f"  Metadata: {len(metadata_df):,} books")
print(f"  Reviews: {len(reviews_df):,} reviews")    

2025-11-09 15:40:22,226 - data_loader - INFO - Joining metadata and reviews with 'inner' join...
2025-11-09 15:40:22,704 - data_loader - INFO - Filtered to books with >=5 reviews: 13,612 books, 714,864 reviews
2025-11-09 15:40:24,093 - data_loader - INFO - Join complete: 714,864 rows (20,000 books × 714,864 reviews)
2025-11-09 15:40:24,737 - data_loader - INFO - Join statistics: 13,612 unique books, 584,538 unique users
2025-11-09 15:40:25,081 - data_loader - INFO - Analyzing rating distribution by category_level_2_sub...
2025-11-09 15:40:25,091 - data_loader - INFO - Joining metadata and reviews with 'inner' join...




Joined data shape:
 (714864, 39)


2025-11-09 15:40:27,183 - data_loader - INFO - Join complete: 727,876 rows (20,000 books × 727,876 reviews)
2025-11-09 15:40:27,807 - data_loader - INFO - Join statistics: 19,991 unique books, 593,235 unique users
2025-11-09 15:40:28,021 - data_loader - INFO - Analyzed 50 categories
2025-11-09 15:40:28,173 - data_loader - INFO - Exporting summary report to: ./artifacts_v2/datasets/complete_data_report.txt
2025-11-09 15:40:28,176 - data_loader - INFO - Generating data summary...
2025-11-09 15:40:28,351 - data_loader - INFO - Data summary generated successfully
2025-11-09 15:40:28,352 - data_loader - INFO - Checking missing values...




Rating distribution by category_level_2_sub category:

                category_level_2_sub  num_reviews  avg_rating  rating_std  \
29              Literature & Fiction       157693    4.266974    1.149147   
31      Mystery, Thriller & Suspense       129952    4.224437    1.123387   
10                  Children's Books       101017    4.638279    0.908919   
46                Teen & Young Adult        35216    4.460047    0.982247   
3              Biographies & Memoirs        32847    4.454166    1.038777   
11          Christian Books & Bibles        27946    4.673549    0.756218   
14            Cookbooks, Food & Wine        27807    4.539900    0.989324   
15            Crafts, Hobbies & Home        20495    4.581800    0.916932   
1                 Arts & Photography        18981    4.546388    0.964556   
34        Politics & Social Sciences        18217    4.458747    1.058231   
23                           History        17443    4.474918    0.978902   
8                  

2025-11-09 15:40:28,392 - data_loader - WARNING - Found missing values in required columns:
                  column  missing_count  missing_percentage   dtype
13  category_level_2_sub            118                0.59  object
2025-11-09 15:40:28,397 - data_loader - INFO - Analyzing category distribution for: category_level_2_sub
2025-11-09 15:40:28,399 - data_loader - INFO - Found 20 unique categories at level 'category_level_2_sub'
2025-11-09 15:40:28,401 - data_loader - INFO - Generating reviews summary...
2025-11-09 15:40:29,244 - data_loader - INFO - Reviews summary generated successfully
2025-11-09 15:40:29,245 - data_loader - INFO - Analyzing reviews per book...
2025-11-09 15:40:29,646 - data_loader - INFO - Analyzed 19,991 books
2025-11-09 15:40:29,661 - data_loader - INFO - Report exported successfully to: ./artifacts_v2/datasets/complete_data_report.txt



Loaded:
  Metadata: 20,000 books
  Reviews: 727,876 reviews


In [11]:
import importlib
import data_preprocessor
importlib.reload(data_preprocessor)

import constants_v2
importlib.reload(constants_v2)
from data_preprocessor import DataPreprocessor

from constants_v2 import PATH_ARTIFACTS_V2, SAMPLE_METADATA_FILE_V2, CLASS_WEIGHTS_FILE_V2

In [47]:

# ========================================
# STEP 2: INITIALIZE PREPROCESSOR
# ========================================
print("\n" + "="*70)
print("STEP 2: INITIALIZING PREPROCESSOR")
print("="*70)

#decrease categories choosen 
#min_samples_per_category 
# Step 2: Preprocess
preprocessor = DataPreprocessor(
        min_samples_per_category=100,    # Minimum 100 books per category
        sampling_percentage=0.5,         # Use 20% of data (for faster training)
        balance_strategy='class_weight', # Calculate weights for imbalanced classes
        min_reviews_per_book=5          # Minimum 5 reviews per book
)

2025-11-09 15:51:41,263 - data_preprocessor - INFO - DataPreprocessor initialized
2025-11-09 15:51:41,263 - data_preprocessor - INFO -   Min samples per category: 100
2025-11-09 15:51:41,264 - data_preprocessor - INFO -   Sampling percentage: 50.0%
2025-11-09 15:51:41,265 - data_preprocessor - INFO -   Balance strategy: class_weight
2025-11-09 15:51:41,266 - data_preprocessor - INFO -   Min reviews per book: 5



STEP 2: INITIALIZING PREPROCESSOR


In [48]:
# ========================================
# STEP 2.1: CLEAN METADATA
# ========================================
print("\n" + "="*70)
print("STEP 2.1: CLEANING METADATA")
print("="*70)
    
# Clean both datasets
clean_metadata = preprocessor.clean_metadata(metadata_df)



2025-11-09 15:51:44,437 - data_preprocessor - INFO - ======================================================================
2025-11-09 15:51:44,438 - data_preprocessor - INFO - CLEANING METADATA
2025-11-09 15:51:44,439 - data_preprocessor - INFO - ======================================================================
2025-11-09 15:51:44,450 - data_preprocessor - INFO - 
[1/5] Removing duplicates...
2025-11-09 15:51:44,473 - data_preprocessor - INFO -   Removed 0 duplicates
2025-11-09 15:51:44,474 - data_preprocessor - INFO - 
[2/5] Handling missing descriptions...



STEP 2.1: CLEANING METADATA


2025-11-09 15:51:44,926 - data_preprocessor - INFO -   Missing descriptions before: 0
2025-11-09 15:51:44,927 - data_preprocessor - INFO -   Missing descriptions after: 0
2025-11-09 15:51:44,928 - data_preprocessor - INFO -   Filled: 0
2025-11-09 15:51:44,929 - data_preprocessor - INFO - 
[3/5] Cleaning text fields...
2025-11-09 15:51:44,929 - data_preprocessor - INFO - Cleaning text field with 20000 entries...
2025-11-09 15:51:45,027 - data_preprocessor - INFO - Text cleaning complete:
2025-11-09 15:51:45,028 - data_preprocessor - INFO -   Non-empty before: 20000
2025-11-09 15:51:45,029 - data_preprocessor - INFO -   Non-empty after: 20000
2025-11-09 15:51:45,031 - data_preprocessor - INFO - Cleaning text field with 20000 entries...
2025-11-09 15:51:48,478 - data_preprocessor - INFO - Text cleaning complete:
2025-11-09 15:51:48,479 - data_preprocessor - INFO -   Non-empty before: 20000
2025-11-09 15:51:48,480 - data_preprocessor - INFO -   Non-empty after: 20000
2025-11-09 15:51:48,48

In [49]:

# ========================================
# STEP 2.2: CLEAN REVIEWS
# ========================================
print("\n" + "="*70)
print("STEP 2.2: CLEANING REVIEWS")
print("="*70)
    
clean_reviews = preprocessor.clean_reviews(reviews_df)    


2025-11-09 15:51:52,308 - data_preprocessor - INFO - 
2025-11-09 15:51:52,309 - data_preprocessor - INFO - CLEANING REVIEWS
2025-11-09 15:51:52,309 - data_preprocessor - INFO - ======================================================================
2025-11-09 15:51:52,380 - data_preprocessor - INFO - 
[1/5] Removing duplicates...



STEP 2.2: CLEANING REVIEWS


2025-11-09 15:51:53,621 - data_preprocessor - INFO -   Removed 7892 duplicate reviews
2025-11-09 15:51:53,622 - data_preprocessor - INFO - 
[2/5] Validating ratings...
2025-11-09 15:51:53,773 - data_preprocessor - INFO -   Removed 0 reviews with invalid ratings
2025-11-09 15:51:53,775 - data_preprocessor - INFO - 
[3/5] Cleaning review text...
2025-11-09 15:51:53,776 - data_preprocessor - INFO - Cleaning text field with 719984 entries...
2025-11-09 15:52:05,930 - data_preprocessor - INFO - Text cleaning complete:
2025-11-09 15:52:05,932 - data_preprocessor - INFO -   Non-empty before: 719874
2025-11-09 15:52:05,933 - data_preprocessor - INFO -   Non-empty after: 719287
2025-11-09 15:52:06,263 - data_preprocessor - INFO -   Removed 697 reviews with empty text
2025-11-09 15:52:06,265 - data_preprocessor - INFO - 
[4/5] Processing verified_purchase field...
2025-11-09 15:52:06,270 - data_preprocessor - INFO -   Verified purchases: 517,441 (71.9%)
2025-11-09 15:52:06,271 - data_preprocesso

In [50]:
# ========================================
# STEP 2.3: ENRICH METADATA WITH REVIEWS
# ========================================
# Enrich metadata with review features
enriched_metadata = preprocessor.enrich_with_reviews(clean_metadata, clean_reviews)


2025-11-09 15:52:14,734 - data_preprocessor - INFO - 
2025-11-09 15:52:14,735 - data_preprocessor - INFO - ENRICHING METADATA WITH REVIEW FEATURES
2025-11-09 15:52:14,736 - data_preprocessor - INFO - ======================================================================
2025-11-09 15:52:14,736 - data_preprocessor - INFO - 
Input data:
2025-11-09 15:52:14,737 - data_preprocessor - INFO -   Metadata books: 19,882
2025-11-09 15:52:14,737 - data_preprocessor - INFO -   Total reviews: 719,287
2025-11-09 15:52:15,017 - data_preprocessor - INFO - 
Aggregating review features...
2025-11-09 15:52:17,194 - data_preprocessor - INFO - 
Aggregated features for 19,987 books
2025-11-09 15:52:17,194 - data_preprocessor - INFO - Features created: ['parent_asin', 'review_count', 'avg_rating_from_reviews', 'rating_std', 'verified_purchase_pct', 'helpful_votes_total', 'review_text_avg_length']
2025-11-09 15:52:17,195 - data_preprocessor - INFO - 
Merging with metadata...
2025-11-09 15:52:17,242 - data_pre

In [51]:

# ========================================
# STEP 2.4: FILTER RARE CATEGORIES (categories with less than 50)
# ========================================    
# Filter and sample
filtered_metadata  = preprocessor.filter_rare_categories(enriched_metadata,
        category_column=TARGET_COLUMN_CATEGORY_LEVEL2)

print(f"\nFiltered metadata:")
print(f"  Books: {len(filtered_metadata):,}")
print(f"  Categories: {filtered_metadata[TARGET_COLUMN_CATEGORY_LEVEL2].nunique()}")


2025-11-09 15:52:21,343 - data_preprocessor - INFO - 
2025-11-09 15:52:21,343 - data_preprocessor - INFO - FILTERING RARE CATEGORIES
2025-11-09 15:52:21,344 - data_preprocessor - INFO - ======================================================================
2025-11-09 15:52:21,347 - data_preprocessor - INFO - 
Before filtering:
2025-11-09 15:52:21,348 - data_preprocessor - INFO -   Total books: 19,882
2025-11-09 15:52:21,349 - data_preprocessor - INFO -   Total categories: 50
2025-11-09 15:52:21,385 - data_preprocessor - INFO - 
After filtering (min 100 samples):
2025-11-09 15:52:21,386 - data_preprocessor - INFO -   Total books: 19,612
2025-11-09 15:52:21,388 - data_preprocessor - INFO -   Total categories: 28
2025-11-09 15:52:21,389 - data_preprocessor - INFO -   Removed categories: 22
2025-11-09 15:52:21,390 - data_preprocessor - INFO -   Removed books: 270
2025-11-09 15:52:21,393 - data_preprocessor - INFO - 
Removed categories: ['Large Print', 'LGBTQ+ Books', 'Popular Suspense', 'B


Filtered metadata:
  Books: 19,612
  Categories: 28


In [52]:
   
# ========================================
# STEP 2.5: FILTER BY REVIEW COUNT (minimal reviews)
# ========================================
print("\n" + "="*70)
print("STEP 2.5: FILTERING BY REVIEW COUNT")
print("="*70)
    
filtered_metadata = preprocessor.filter_by_review_count( filtered_metadata)

2025-11-09 15:52:27,429 - data_preprocessor - INFO - 
2025-11-09 15:52:27,431 - data_preprocessor - INFO - FILTERING BY REVIEW COUNT
2025-11-09 15:52:27,433 - data_preprocessor - INFO - ======================================================================
2025-11-09 15:52:27,434 - data_preprocessor - INFO - 
Filtering books with less than 5 reviews...
2025-11-09 15:52:27,456 - data_preprocessor - INFO -   Books before: 19,612
2025-11-09 15:52:27,457 - data_preprocessor - INFO -   Books after: 13,360
2025-11-09 15:52:27,458 - data_preprocessor - INFO -   Removed: 6,252 books



STEP 2.5: FILTERING BY REVIEW COUNT


In [53]:
# ========================================
# STEP 2.6: STRATIFIED SAMPLING
# ========================================
print("\n" + "="*70)
print("STEP 2.6: STRATIFIED SAMPLING")
print("="*70)

sampled_metadata  = preprocessor.sample_data(filtered_metadata,
        category_column=TARGET_COLUMN_CATEGORY_LEVEL2,
        random_state=42)

print(f"\nSampled metadata:")
print(f"  Books: {len(sampled_metadata):,}")
print(f"  Categories: {sampled_metadata[TARGET_COLUMN_CATEGORY_LEVEL2].nunique()}")


2025-11-09 15:52:35,796 - data_preprocessor - INFO - 
2025-11-09 15:52:35,797 - data_preprocessor - INFO - STRATIFIED SAMPLING
2025-11-09 15:52:35,797 - data_preprocessor - INFO - ======================================================================
2025-11-09 15:52:35,798 - data_preprocessor - INFO - 
Sampling 50.0% of data
2025-11-09 15:52:35,798 - data_preprocessor - INFO -   Original: 13,360 books
2025-11-09 15:52:35,799 - data_preprocessor - INFO -   Target: 6,680 books



STEP 2.6: STRATIFIED SAMPLING


d:\workspace\datatalks-club\DataTalksClub\mid-term-project\data_preprocessor.py:605: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df_sampled = df.groupby(category_column, group_keys=False).apply(
2025-11-09 15:52:36,385 - data_preprocessor - INFO -   Sampled: 6,681 books
2025-11-09 15:52:36,387 - data_preprocessor - INFO - 
Verifying category distribution:
2025-11-09 15:52:36,393 - data_preprocessor - INFO -   Literature & Fiction: 24.8% → 24.8%
2025-11-09 15:52:36,394 - data_preprocessor - INFO -   Mystery, Thriller & Suspense: 12.9% → 12.9%
2025-11-09 15:52:36,395 - data_preprocessor - INFO -   Children's Books: 11.1% → 11.1%
2025-11-09 15:52:36,396 - data_preprocessor - INFO -   Christian Books & 


Sampled metadata:
  Books: 6,681
  Categories: 28


In [54]:
# ========================================
# STEP 2.7: CALCULATE CLASS WEIGHTS
# ========================================
    
# Calculate class weights
class_weights = preprocessor.calculate_class_weights(sampled_metadata, 
                                                      category_column=TARGET_COLUMN_CATEGORY_LEVEL2)

      


2025-11-09 15:52:46,783 - data_preprocessor - INFO - 
2025-11-09 15:52:46,784 - data_preprocessor - INFO - CALCULATING CLASS WEIGHTS
2025-11-09 15:52:46,785 - data_preprocessor - INFO - ======================================================================
2025-11-09 15:52:48,407 - data_preprocessor - INFO - 
Balance strategy: class_weight
2025-11-09 15:52:48,408 - data_preprocessor - INFO - Computed weights for 28 categories
2025-11-09 15:52:48,409 - data_preprocessor - INFO - 
Top 5 minority classes (highest weights):
2025-11-09 15:52:48,411 - data_preprocessor - INFO -   Education & Teaching: weight=6.63, samples=36
2025-11-09 15:52:48,413 - data_preprocessor - INFO -   Medical Books: weight=6.63, samples=36
2025-11-09 15:52:48,415 - data_preprocessor - INFO -   Computers & Technology: weight=5.08, samples=47
2025-11-09 15:52:48,416 - data_preprocessor - INFO -   Travel: weight=4.68, samples=51
2025-11-09 15:52:48,416 - data_preprocessor - INFO -   Engineering & Transportation: weig

In [55]:
# ========================================
# STEP 2.8: SAVE CLASS WEIGHTS  AND SAMPLED DATA
# ========================================
preprocessor.save_preprocessed_data(sampled_metadata,
        output_path=SAMPLE_METADATA_FILE_V2,
        save_class_weights=True,
        class_weights=class_weights,
        weights_path=CLASS_WEIGHTS_FILE_V2)

✓ Directory ensured: artifacts_v2\dataset
✓ CSV saved: ./artifacts_v2//dataset/sampled_metadata.csv
✓ Directory ensured: artifacts_v2\dataset
✓ Class weights saved: ./artifacts_v2//dataset/class_weights.json


In [56]:
# ========================================
# FINAL SUMMARY
# ========================================    
# Print summary
preprocessor.print_preprocessing_summary()
    
print("\n" + "="*70)
print("PREPROCESSING COMPLETE")
print("="*70)
    
print(f"\nFinal dataset statistics:")
print(f"  Total books: {len(sampled_metadata):,}")
print(f"  Total categories: {sampled_metadata[TARGET_COLUMN_CATEGORY_LEVEL2].nunique()}")
print(f"  Total columns: {len(sampled_metadata.columns)}")
print(f"  Avg reviews per book: {sampled_metadata['review_count'].mean():.1f}")
print(f"  Avg rating: {sampled_metadata['avg_rating_from_reviews'].mean():.2f}")
    
print("\nTop 5 categories by count:")
top_categories = sampled_metadata[TARGET_COLUMN_CATEGORY_LEVEL2].value_counts().head(5)
for cat, count in top_categories.items():
        pct = (count / len(sampled_metadata)) * 100
        print(f"  {cat}: {count:,} books ({pct:.1f}%)")



PREPROCESSING SUMMARY

📚 METADATA CLEANING:
  Original books: 20,000
  Final books: 19,882
  Duplicates removed: 0
  Descriptions filled: 0
  No category removed: 118

💬 REVIEWS CLEANING:
  Original reviews: 727,876
  Final reviews: 719,287
  Duplicates removed: 7892
  Invalid ratings removed: 0
  Empty text removed: 697

⭐ ENRICHMENT:
  Books with reviews: 19,987
  Books without reviews: 13
  Features added: 6

🔍 FILTERING:
  Original books: 19,882
  Final books: 19,612
  Original categories: 50
  Final categories: 28
  Removed categories: 22

📊 SAMPLING:
  Original: 13,360
  Sampled: 6,681
  Percentage: 50%


PREPROCESSING COMPLETE

Final dataset statistics:
  Total books: 6,681
  Total categories: 28
  Total columns: 34
  Avg reviews per book: 50.7
  Avg rating: 4.39

Top 5 categories by count:
  Literature & Fiction: 1,655 books (24.8%)
  Mystery, Thriller & Suspense: 860 books (12.9%)
  Children's Books: 744 books (11.1%)
  Christian Books & Bibles: 362 books (5.4%)
  Biographies

In [57]:
import importlib
import data_splitter
importlib.reload(data_splitter)
from data_splitter import DataSplitter
from constants import TARGET_COLUMN_CATEGORY_LEVEL2
from constants_v2 import TRAIN_FILE_V2, TEST_FILE_V2, INFO_FILE_V2
from constants_v2 import SAMPLE_METADATA_FILE_V2  , PATH_ARTIFACTS_V2
print("\n" + "=" * 80)
print("DATA SPLITTER ")
print("=" * 80)
    
# Inicializar splitter
splitter = DataSplitter(
        test_size=0.2,           # 20% para test
        random_state=42,         # Reproducibilidad
        target_column=TARGET_COLUMN_CATEGORY_LEVEL2,
        train_file_name=TRAIN_FILE_V2,
        test_file_name=TEST_FILE_V2,
        info_file_name=INFO_FILE_V2
)
    

2025-11-09 15:54:11,379 - data_splitter - INFO - DataSplitter initialized
2025-11-09 15:54:11,380 - data_splitter - INFO -   test_size: 0.2
2025-11-09 15:54:11,381 - data_splitter - INFO -   random_state: 42
2025-11-09 15:54:11,381 - data_splitter - INFO -   target_column: category_level_2_sub



DATA SPLITTER 


In [58]:
# Ejecutar pipeline completo
results = splitter.run_pipeline(
        sampled_metadata_path=SAMPLE_METADATA_FILE_V2,
        output_dir=PATH_ARTIFACTS_V2
    )
    
print("\n✅ Data splitting completed!")
print(f"\nAccess files in: {PATH_ARTIFACTS_V2}/")
print("  - train_metadata.csv")
print("  - test_metadata.csv")
print("  - split_info.json")

2025-11-09 15:54:14,703 - data_splitter - INFO - 
2025-11-09 15:54:14,704 - data_splitter - INFO - RUNNING DATA SPLITTER PIPELINE
2025-11-09 15:54:14,705 - data_splitter - INFO - ================================================================================
2025-11-09 15:54:14,705 - data_splitter - INFO - ================================================================================
2025-11-09 15:54:14,706 - data_splitter - INFO - LOADING SAMPLED METADATA
2025-11-09 15:54:14,706 - data_splitter - INFO - ================================================================================
2025-11-09 15:54:14,707 - data_splitter - INFO - Loading: ./artifacts_v2//dataset/sampled_metadata.csv
2025-11-09 15:54:15,778 - data_splitter - INFO - ✓ Data loaded successfully
2025-11-09 15:54:15,780 - data_splitter - INFO -   Shape: (6681, 34)
2025-11-09 15:54:15,781 - data_splitter - INFO -   Columns: ['title', 'subtitle', 'author_name', 'author_about', 'publisher', 'publisher_date', 'format', 'pag


✅ Data splitting completed!

Access files in: ./artifacts_v2//
  - train_metadata.csv
  - test_metadata.csv
  - split_info.json


In [59]:

import pandas as pd
from constants_v2 import PATH_ARTIFACTS_V2, TRAIN_FILE_V2, TEST_FILE_V2
df_train = pd.read_csv(f'{PATH_ARTIFACTS_V2}{TRAIN_FILE_V2}')
df_test = pd.read_csv(f'{PATH_ARTIFACTS_V2}{TEST_FILE_V2}')

print(f"\nTrain columns: {df_train.columns}")
print(f"Test columns: {df_test.columns}")


Train columns: Index(['title', 'subtitle', 'author_name', 'author_about', 'publisher',
       'publisher_date', 'format', 'page_count', 'language', 'isbn_10',
       'isbn_13', 'main_category', 'category_level_1_main',
       'category_level_2_sub', 'category_level_3_detail', 'average_rating',
       'rating_number', 'price', 'price_numeric', 'description',
       'features_text', 'dimensions', 'item_weight', 'images', 'videos',
       'store', 'parent_asin', 'bought_together', 'review_count',
       'avg_rating_from_reviews', 'rating_std', 'verified_purchase_pct',
       'helpful_votes_total', 'review_text_avg_length'],
      dtype='object')
Test columns: Index(['title', 'subtitle', 'author_name', 'author_about', 'publisher',
       'publisher_date', 'format', 'page_count', 'language', 'isbn_10',
       'isbn_13', 'main_category', 'category_level_1_main',
       'category_level_2_sub', 'category_level_3_detail', 'average_rating',
       'rating_number', 'price', 'price_numeric', 'des

In [60]:
import importlib
import feature_extractor
importlib.reload(feature_extractor)
from feature_extractor import FeatureExtractor
import pandas as pd

# Apply FeatureExtractor
extractor = FeatureExtractor(
        max_tfidf_features=8000,
        tfidf_ngram_range=(1, 3)
    )

2025-11-09 15:54:36,139 - feature_extractor - INFO - FeatureExtractor initialized successfully
2025-11-09 15:54:36,140 - feature_extractor - INFO - Configuration: {'max_tfidf_features': 8000, 'tfidf_ngram_range': (1, 3), 'tfidf_min_df': 2, 'tfidf_max_df': 0.95}


In [61]:
#. Fit  train (fits TF-IDF, Scaler, LabelEncoders and Target Encoder)
# Select features numerical, categorical and engineered. also vectorized features
# from title and description columns
extractor.fit(df_train, target_column=TARGET_COLUMN_CATEGORY_LEVEL2)


2025-11-09 15:54:40,144 - feature_extractor - INFO - ================================================================================
2025-11-09 15:54:40,145 - feature_extractor - INFO - FITTING FEATURE EXTRACTOR
2025-11-09 15:54:40,145 - feature_extractor - INFO - ================================================================================
2025-11-09 15:54:40,146 - feature_extractor - INFO - Input shape: (5344, 34)
2025-11-09 15:54:40,147 - feature_extractor - INFO - 
0. Fitting LabelEncoder for target...
2025-11-09 15:54:40,149 - feature_extractor - INFO -    ✓ Target classes: 28
2025-11-09 15:54:40,149 - feature_extractor - INFO -       0: Arts & Photography (146 samples)
2025-11-09 15:54:40,150 - feature_extractor - INFO -       1: Biographies & Memoirs (280 samples)
2025-11-09 15:54:40,151 - feature_extractor - INFO -       2: Business & Money (123 samples)
2025-11-09 15:54:40,152 - feature_extractor - INFO -       3: Children's Books (595 samples)
2025-11-09 15:54:40,153 - fe

In [62]:

import numpy as np
print("FeatureExtractor - Book Classification Pipeline")

# Transform train and test (returns X and y)
X_train, y_train, features_train, feature_names = extractor.transform(df_train, target_column=TARGET_COLUMN_CATEGORY_LEVEL2)
X_test, y_test, features_test, _ = extractor.transform(df_test, target_column=TARGET_COLUMN_CATEGORY_LEVEL2)    

2025-11-09 15:56:52,044 - feature_extractor - INFO - ================================================================================
2025-11-09 15:56:52,045 - feature_extractor - INFO - TRANSFORMING FEATURES
2025-11-09 15:56:52,046 - feature_extractor - INFO - ================================================================================
2025-11-09 15:56:52,046 - feature_extractor - INFO - Input shape: (5344, 34)
2025-11-09 15:56:52,047 - feature_extractor - INFO - 
0. Extracting and encoding target...
2025-11-09 15:56:52,051 - feature_extractor - INFO -    ✓ Target shape: (5344,)
2025-11-09 15:56:52,053 - feature_extractor - INFO -    Target distribution:
2025-11-09 15:56:52,054 - feature_extractor - INFO -       0: Arts & Photography             -   146 (  2.7%)
2025-11-09 15:56:52,056 - feature_extractor - INFO -       1: Biographies & Memoirs          -   280 (  5.2%)
2025-11-09 15:56:52,057 - feature_extractor - INFO -       2: Business & Money               -   123 (  2.3%)
20

FeatureExtractor - Book Classification Pipeline


2025-11-09 15:57:01,892 - feature_extractor - INFO -    ✓ Text features shape: (5344, 8000)
2025-11-09 15:57:01,894 - feature_extractor - INFO - 
2. Transforming numerical features...
2025-11-09 15:57:01,907 - feature_extractor - INFO -    ✓ Numerical features shape: (5344, 6)
2025-11-09 15:57:01,908 - feature_extractor - INFO - 
3. Transforming categorical features...
2025-11-09 15:57:01,957 - feature_extractor - INFO -    ✓ Categorical features shape: (5344, 3)
2025-11-09 15:57:01,960 - feature_extractor - INFO - 
4. Extracting engineered features...
2025-11-09 15:57:02,018 - feature_extractor - INFO -    ✓ Engineered features shape: (5344, 5)
2025-11-09 15:57:02,020 - feature_extractor - INFO - 
5. Concatenating all features...
2025-11-09 15:57:02,220 - feature_extractor - INFO -    ✓ Final feature matrix shape: (5344, 8014)
2025-11-09 15:57:02,223 - feature_extractor - INFO - 
2025-11-09 15:57:02,224 - feature_extractor - INFO - FEATURE TRANSFORMATION COMPLETED
2025-11-09 15:57:02,

In [64]:
import importlib
import feature_extractor_serializer
importlib.reload(feature_extractor_serializer)
from feature_extractor_serializer import save_features_and_metadata
from constants_v2 import FEATURE_EXTRACTOR_FEATURES_OUTPUT_DIR_V2, MODEL_OUTPUT_DIR_V2

print(f"\nExtractor stats: \n")
print("="*80)
print(f'{extractor.get_stats()}')
#Save for later use

save_features_and_metadata(
        X_train, y_train,
        X_test, y_test,
        feature_names,
        extractor,
        output_dir=FEATURE_EXTRACTOR_FEATURES_OUTPUT_DIR_V2,
        model_dir=MODEL_OUTPUT_DIR_V2)

# After executing  FeatureExtractor obtain the following variables:
'''
X_train         # numpy array (N_train, 5014)
y_train         # numpy array (N_train,)
X_test          # numpy array (N_test, 5014)
y_test          # numpy array (N_test,)
feature_names   # list with 5014 feature names
'''

### **Detailed Features  (5014 total):**
'''
📊 Features:
├── Text (TF-IDF):    5000 features
│   └── tfidf_0, tfidf_1, ..., tfidf_4999
│
├── Numerical:        7 features
│   ├── price
│   ├── average_rating
│   ├── review_count
│   ├── rating_std
│   ├── verified_purchase_pct
│   ├── helpful_votes_total
│   └── avg_review_length
│
├── Categorical:      3 features (encoded)
│   ├── format
│   ├── binned_rating
│   └── price_category
│
└── Engineered:       5 features
    ├── title_length
    ├── desc_length
    ├── has_series
    ├── price_rating_ratio
    └── engagement_score
'''

#loaded_extractor = FeatureExtractor.load('models/feature_extractor')

2025-11-09 15:57:29,781 - feature_extractor_serializer - INFO - ================================================================================
2025-11-09 15:57:29,782 - feature_extractor_serializer - INFO - SAVING ALL OUTPUTS
2025-11-09 15:57:29,783 - feature_extractor_serializer - INFO - ================================================================================
2025-11-09 15:57:29,785 - feature_extractor_serializer - INFO - 
1. Saving numpy arrays...



Extractor stats: 

{'status': 'fitted', 'total_features': 8014, 'feature_breakdown': {'text': 8000, 'numerical': 6, 'categorical': 3, 'engineered': 5}, 'target_info': {'n_classes': 28, 'classes': ['Arts & Photography', 'Biographies & Memoirs', 'Business & Money', "Children's Books", 'Christian Books & Bibles', 'Comics & Graphic Novels', 'Computers & Technology', 'Cookbooks, Food & Wine', 'Crafts, Hobbies & Home', 'Education & Teaching', 'Engineering & Transportation', 'Health, Fitness & Dieting', 'History', 'Humor & Entertainment', 'Literature & Fiction', 'Medical Books', 'Mystery, Thriller & Suspense', 'New, Used & Rental Textbooks', 'Politics & Social Sciences', 'Reference', 'Religion & Spirituality', 'Romance', 'Science & Math', 'Science Fiction & Fantasy', 'Self-Help', 'Sports & Outdoors', 'Teen & Young Adult', 'Travel']}, 'config': {'max_tfidf_features': 8000, 'tfidf_ngram_range': (1, 3), 'tfidf_min_df': 2, 'tfidf_max_df': 0.95}, 'feature_names': {'text': ['tfidf_0', 'tfidf_1', '

2025-11-09 15:57:30,238 - feature_extractor_serializer - INFO -    ✓ X_train saved: (5344, 8014)
2025-11-09 15:57:30,240 - feature_extractor_serializer - INFO -    ✓ y_train saved: (5344,)
2025-11-09 15:57:30,362 - feature_extractor_serializer - INFO -    ✓ X_test saved: (1337, 8014)
2025-11-09 15:57:30,364 - feature_extractor_serializer - INFO -    ✓ y_test saved: (1337,)
2025-11-09 15:57:30,364 - feature_extractor_serializer - INFO - 
2. Saving feature names...
2025-11-09 15:57:30,378 - feature_extractor_serializer - INFO -    ✓ feature_names.json saved: 8014 features
2025-11-09 15:57:30,380 - feature_extractor_serializer - INFO - 
3. Saving class names and mapping...
2025-11-09 15:57:30,384 - feature_extractor_serializer - INFO -    ✓ class_names.json saved: 28 classes
2025-11-09 15:57:30,385 - feature_extractor_serializer - INFO - 
4. Saving FeatureExtractor...
2025-11-09 15:57:30,387 - feature_extractor - INFO - 
Saving FeatureExtractor to ./artifacts_v2//models/...
2025-11-09 15:

'\n📊 Features:\n├── Text (TF-IDF):    5000 features\n│   └── tfidf_0, tfidf_1, ..., tfidf_4999\n│\n├── Numerical:        7 features\n│   ├── price\n│   ├── average_rating\n│   ├── review_count\n│   ├── rating_std\n│   ├── verified_purchase_pct\n│   ├── helpful_votes_total\n│   └── avg_review_length\n│\n├── Categorical:      3 features (encoded)\n│   ├── format\n│   ├── binned_rating\n│   └── price_category\n│\n└── Engineered:       5 features\n    ├── title_length\n    ├── desc_length\n    ├── has_series\n    ├── price_rating_ratio\n    └── engagement_score\n'

In [65]:
import numpy as np
import json

from constants_v2 import X_TRAIN_FILE_V2, Y_TRAIN_FILE_V2, X_TEST_FILE_V2, Y_TEST_FILE_V2
from constants_v2 import CLASS_NAMES_FILE_V2, CLASS_WEIGHTS_FILE_V2,FEATURE_NAMES_FILE_V2
from constants_v2 import REPORT_XGBOOST_CLASSIFICATION_REPORT_FILE_V2, REPORT_XGBOOST_FEATURE_IMPORTANCE_FILE_V2
from constants_v2 import REPORT_XGBOOST_METRICS_FILE_V2,REPORT_XGBOOST_MODEL_FILE_V2, REPORT_XGBOOST_CONFUSION_MATRIX_FILE_V2

print("\n" + "=" * 80)
print("STEP 4: MODEL TRAINING")
print("=" * 80)

# ------------------------------------------------------------------------------
# 1. LOAD DATA FROM STEP 3
# ------------------------------------------------------------------------------
print("Loading data ")

X_train = np.load(X_TRAIN_FILE_V2)
y_train = np.load(Y_TRAIN_FILE_V2)
X_test = np.load(X_TEST_FILE_V2)
y_test = np.load(Y_TEST_FILE_V2)

# Load class names
with open(CLASS_NAMES_FILE_V2, 'r') as f:
    class_info = json.load(f)
    class_names = class_info['class_names']

# Load feature names (optional, for visualizations)
with open(FEATURE_NAMES_FILE_V2, 'r') as f:
    feature_info = json.load(f)
    feature_names = feature_info['feature_names']

# Load class weights (if available from Step 2)
try:
    with open(CLASS_WEIGHTS_FILE_V2, 'r') as f:
        class_weights = json.load(f)
except FileNotFoundError:
    class_weights = None
    print("No class_weights.json found, will use default balancing")

print(f"Data loaded:")
print(f"  Training samples: {len(X_train)}")
print(f"  Test samples: {len(X_test)}")
print(f"  Features: {X_train.shape[1]}")
print(f"  Classes: {len(class_names)}")




STEP 4: MODEL TRAINING
Loading data 
Data loaded:
  Training samples: 5344
  Test samples: 1337
  Features: 8014
  Classes: 28


In [66]:
import importlib
import model_trainer
importlib.reload(model_trainer)
from model_trainer import XGBoostTrainer
# ------------------------------------------------------------------------------
# 2. TRAIN XGBOOST MODEL
# ------------------------------------------------------------------------------
print("\\n" + "="*80)
print("Training XGBoost Model...")
print("="*80)

xgb_trainer = XGBoostTrainer(
    X_train=X_train,
    y_train=y_train,
    X_test=X_test,
    y_test=y_test,
    class_names=class_names,
    class_weights=class_weights,
    output_dir='models',
    n_estimators=200,
    max_depth=8,
    learning_rate=0.1,
    subsample=0.8,
    colsample_bytree=0.8,
    early_stopping_rounds=10
)

2025-11-09 15:57:59,594 - model_trainer - INFO - Initialized XGBoost Trainer
2025-11-09 15:57:59,595 - model_trainer - INFO - Training samples: 5344, Test samples: 1337
2025-11-09 15:57:59,596 - model_trainer - INFO - Number of features: 8014
2025-11-09 15:57:59,598 - model_trainer - INFO - Number of classes: 28


\n================================================================================
Training XGBoost Model...


In [67]:
# Train
xgb_trainer.train()

2025-11-09 15:58:03,008 - model_trainer - INFO - Training XGBoost with parameters: {'n_estimators': 200, 'max_depth': 8, 'learning_rate': 0.1, 'subsample': 0.8, 'colsample_bytree': 0.8, 'early_stopping_rounds': 10}
2025-11-09 15:58:03,010 - model_trainer - WARNING - Error computing sample weights: invalid literal for int() with base 10: 'Arts & Photography'
2025-11-09 15:58:03,010 - model_trainer - INFO - Falling back to no sample weights
2025-11-09 15:58:03,011 - model_trainer - INFO - Attempting to use early stopping with 10 rounds
2025-11-09 15:58:03,012 - model_trainer - WARNING - XGBoost version doesn't support early_stopping_rounds in fit(): XGBClassifier.fit() got an unexpected keyword argument 'early_stopping_rounds'
2025-11-09 15:58:03,012 - model_trainer - INFO - Training without early stopping...


[0]	validation_0-mlogloss:2.96824
[1]	validation_0-mlogloss:2.77368
[2]	validation_0-mlogloss:2.61851
[3]	validation_0-mlogloss:2.49837
[4]	validation_0-mlogloss:2.39904
[5]	validation_0-mlogloss:2.31911
[6]	validation_0-mlogloss:2.24507
[7]	validation_0-mlogloss:2.18254
[8]	validation_0-mlogloss:2.12482
[9]	validation_0-mlogloss:2.07531
[10]	validation_0-mlogloss:2.03073
[11]	validation_0-mlogloss:1.98883
[12]	validation_0-mlogloss:1.95306
[13]	validation_0-mlogloss:1.91922
[14]	validation_0-mlogloss:1.88925
[15]	validation_0-mlogloss:1.86219
[16]	validation_0-mlogloss:1.83511
[17]	validation_0-mlogloss:1.81123
[18]	validation_0-mlogloss:1.78655
[19]	validation_0-mlogloss:1.76468
[20]	validation_0-mlogloss:1.74279
[21]	validation_0-mlogloss:1.72362
[22]	validation_0-mlogloss:1.70633
[23]	validation_0-mlogloss:1.69063
[24]	validation_0-mlogloss:1.67371
[25]	validation_0-mlogloss:1.65921
[26]	validation_0-mlogloss:1.64544
[27]	validation_0-mlogloss:1.63359
[28]	validation_0-mlogloss:1.6

2025-11-09 16:27:09,875 - model_trainer - INFO - XGBoost training completed


In [68]:
# Evaluate on test set
xgb_test_metrics = xgb_trainer.evaluate(X_test, y_test, 'test')

2025-11-09 16:27:42,674 - model_trainer - INFO - Evaluating XGBoost on test set...
2025-11-09 16:27:43,037 - model_trainer - INFO - TEST METRICS:
2025-11-09 16:27:43,037 - model_trainer - INFO -   Accuracy: 0.5849
2025-11-09 16:27:43,038 - model_trainer - INFO -   Precision (weighted): 0.5816
2025-11-09 16:27:43,039 - model_trainer - INFO -   Recall (weighted): 0.5849
2025-11-09 16:27:43,039 - model_trainer - INFO -   F1-Score (weighted): 0.5558


In [77]:
# Cross-validation
xgb_cv_results = xgb_trainer.cross_validate(cv=2, scoring='f1_weighted')


2025-11-09 16:45:01,566 - model_trainer - INFO - Performing 2-fold cross-validation...
2025-11-09 17:11:33,443 - model_trainer - INFO - CV f1_weighted: 0.5446 (+/- 0.0003)


In [78]:

import importlib
import constants_v2
importlib.reload(constants_v2)

from constants_v2 import REPORT_XGBOOST_CLASSIFICATION_REPORT_FILE_V2, REPORT_XGBOOST_FEATURE_IMPORTANCE_FILE_V2
from constants_v2 import REPORT_XGBOOST_METRICS_FILE_V2,REPORT_XGBOOST_MODEL_FILE_V2, REPORT_XGBOOST_CONFUSION_MATRIX_FILE_V2

# Visualizations
xgb_trainer.plot_confusion_matrix(
    y_test,
    xgb_trainer.model.predict(X_test),
    normalize=True,
    save_path=REPORT_XGBOOST_CONFUSION_MATRIX_FILE_V2
)

2025-11-09 17:12:09,988 - model_trainer - INFO - Confusion matrix saved to ./artifacts_v2//reports//xgboost_confusion_matrix.png


In [79]:
xgb_trainer.plot_feature_importance(
    top_n=20,
    feature_names=feature_names,
    save_path=REPORT_XGBOOST_FEATURE_IMPORTANCE_FILE_V2
)

2025-11-09 17:12:13,116 - model_trainer - INFO - Feature importance plot saved to ./artifacts_v2//reports//xgboost_feature_importance.png
2025-11-09 17:12:13,117 - model_trainer - INFO - 
Top 20 Features:
2025-11-09 17:12:13,118 - model_trainer - INFO -   1. tfidf_3401: 0.0018
2025-11-09 17:12:13,119 - model_trainer - INFO -   2. tfidf_3733: 0.0018
2025-11-09 17:12:13,120 - model_trainer - INFO -   3. tfidf_6649: 0.0020
2025-11-09 17:12:13,122 - model_trainer - INFO -   4. tfidf_1424: 0.0020
2025-11-09 17:12:13,122 - model_trainer - INFO -   5. tfidf_7649: 0.0020
2025-11-09 17:12:13,123 - model_trainer - INFO -   6. tfidf_5805: 0.0020
2025-11-09 17:12:13,124 - model_trainer - INFO -   7. tfidf_5713: 0.0020
2025-11-09 17:12:13,125 - model_trainer - INFO -   8. tfidf_6192: 0.0021
2025-11-09 17:12:13,126 - model_trainer - INFO -   9. tfidf_7366: 0.0021
2025-11-09 17:12:13,126 - model_trainer - INFO -   10. tfidf_639: 0.0021
2025-11-09 17:12:13,127 - model_trainer - INFO -   11. tfidf_6186

In [80]:
xgb_trainer.save_classification_report(
    y_test,
    xgb_trainer.model.predict(X_test),
    save_path=REPORT_XGBOOST_CLASSIFICATION_REPORT_FILE_V2
)
# Save model
xgb_trainer.save_model(REPORT_XGBOOST_MODEL_FILE_V2)

# Save metrics

with open(REPORT_XGBOOST_METRICS_FILE_V2, 'w') as f:
    json.dump({
        'test_metrics': xgb_test_metrics,
        'cv_results': xgb_cv_results
    }, f, indent=2) 


2025-11-09 17:12:30,534 - model_trainer - INFO - 
XGBoost Classification Report:
                              precision    recall  f1-score   support

          Arts & Photography     0.6538    0.4595    0.5397        37
       Biographies & Memoirs     0.4603    0.4143    0.4361        70
            Business & Money     0.6571    0.7419    0.6970        31
            Children's Books     0.7228    0.8926    0.7988       149
    Christian Books & Bibles     0.6329    0.6944    0.6623        72
     Comics & Graphic Novels     0.8182    0.4091    0.5455        22
      Computers & Technology     1.0000    0.7778    0.8750         9
      Cookbooks, Food & Wine     0.7500    0.8182    0.7826        33
      Crafts, Hobbies & Home     0.4286    0.3000    0.3529        30
        Education & Teaching     0.4000    0.2857    0.3333         7
Engineering & Transportation     1.0000    0.2727    0.4286        11
   Health, Fitness & Dieting     0.2308    0.1364    0.1714        22
        

In [ ]:
import importlib
import book_classifier
importlib.reload(book_classifier)
from book_classifier import example_single_prediction, BookClassifier, example_batch_prediction

import feature_extractor
importlib.reload(feature_extractor)
from feature_extractor import FeatureExtractor

from constants_v2 import REPORT_XGBOOST_MODEL_FILE_V2, MODEL_OUTPUT_DIR_V2

example_single_prediction(REPORT_XGBOOST_MODEL_FILE_V2, MODEL_OUTPUT_DIR_V2)

example_batch_prediction(REPORT_XGBOOST_MODEL_FILE_V2, MODEL_OUTPUT_DIR_V2)

2025-11-10 12:29:17,979 - feature_extractor - INFO - 
Loading FeatureExtractor from ./artifacts_v2//models/...
2025-11-10 12:29:17,984 - feature_extractor - INFO - FeatureExtractor initialized successfully
2025-11-10 12:29:17,985 - feature_extractor - INFO - Configuration: {'max_tfidf_features': 8000, 'tfidf_ngram_range': [1, 3], 'tfidf_min_df': 2, 'tfidf_max_df': 0.95}



EXAMPLE 2: Batch Prediction
Loading model...
✓ Model loaded from ./artifacts_v2//reports//xgboost_model.pkl


2025-11-10 12:29:18,079 - feature_extractor - INFO -    ✓ FeatureExtractor loaded successfully
2025-11-10 12:29:18,081 - feature_extractor - INFO -    Target classes: 28
2025-11-10 12:29:18,143 - feature_extractor - INFO - ✓ Features for inference: (3, 8014)


✓ Feature extractor loaded
✓ Loaded 28 classes

✓ Classifier ready!


Results:

  Python Crash Course
    → Christian Books & Bibles (31.9%)

  The Great Gatsby
    → Literature & Fiction (83.0%)

  Healthy Cooking Made Easy
    → Cookbooks, Food & Wine (97.0%)


In [2]:
!pip install fastapi uvicorn pydantic

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: C:\Python313\python.exe -m pip install --upgrade pip


In [1]:
!python.exe .\test_api_client.py

^C


    ╔══════════════════════════════════════════════════════════════════════════════╗
    ║                    TEST CLIENT - BOOK CLASSIFIER API                         ║
    ╚══════════════════════════════════════════════════════════════════════════════╝
    

Make sure the API is running at http://localhost:8005
If not running, execute first: python api.py


================================================================================
TEST MENU
================================================================================
1. Test: Single book
2. Test: Multiple books
3. Test: Edge cases
4. Interactive mode
5. Speed benchmark
6. Run all tests
7. Exit

Choose an option (1-7): 6

================================================================================
RUNNING ALL TESTS
================================================================================

================================================================================
TEST 1: Single Book Prediction
================================================================================

✓ API Status: running

================================================================================
📚 Book: The Da Vinci Code
================================================================================

🎯 Predicted Category: Literature & Fiction
📊 Confidence: 71.6%

🏆 Top 3 Predictions:
  1. Literature & Fiction                ███████████████████████████████████ 71.6%
  2. Mystery, Thriller & Suspense        ██████ 13.6%
  3. Christian Books & Bibles            ████ 8.6%

================================================================================
TEST 2: Multiple Books Prediction
================================================================================

================================================================================
📚 Book: Python Crash Course
================================================================================

🎯 Predicted Category: Christian Books & Bibles
📊 Confidence: 36.3%

🏆 Top 3 Predictions:
  1. Christian Books & Bibles            ██████████████████ 36.3%
  2. Literature & Fiction                ███████████████ 30.7%
  3. Computers & Technology              █████ 11.2%

================================================================================
📚 Book: The Great Gatsby
================================================================================

🎯 Predicted Category: Literature & Fiction
📊 Confidence: 86.6%

🏆 Top 3 Predictions:
  1. Literature & Fiction                ███████████████████████████████████████████ 86.6%
  2. Christian Books & Bibles            ██ 4.9%
  3. Humor & Entertainment               █ 3.4%

================================================================================
📚 Book: Healthy Cooking Made Easy
================================================================================

🎯 Predicted Category: Cookbooks, Food & Wine
📊 Confidence: 97.1%

🏆 Top 3 Predictions:
  1. Cookbooks, Food & Wine              ████████████████████████████████████████████████ 97.1%
  2. Literature & Fiction                 1.5%
  3. Christian Books & Bibles             0.6%

================================================================================
📚 Book: Harry Potter and the Sorcerer's Stone
================================================================================

🎯 Predicted Category: Children's Books
📊 Confidence: 77.3%

🏆 Top 3 Predictions:
  1. Children's Books                    ██████████████████████████████████████ 77.3%
  2. Teen & Young Adult                  █████ 10.9%
  3. Christian Books & Bibles            ███ 7.6%

================================================================================
📚 Book: Becoming
================================================================================

🎯 Predicted Category: Biographies & Memoirs
📊 Confidence: 40.1%

🏆 Top 3 Predictions:
  1. Biographies & Memoirs               ████████████████████ 40.1%
  2. Children's Books                    █████████ 19.9%
  3. Christian Books & Bibles            █████████ 19.2%

================================================================================
TEST 3: Edge Cases
================================================================================

📌 Test: Very Expensive Book

================================================================================
📚 Book: Expensive Book
================================================================================

🎯 Predicted Category: Literature & Fiction
📊 Confidence: 54.7%

🏆 Top 3 Predictions:
  1. Literature & Fiction                ███████████████████████████ 54.7%
  2. Christian Books & Bibles            ████ 8.3%
  3. Romance                             ███ 7.1%

📌 Test: Very Cheap Book

================================================================================
📚 Book: Cheap Book
================================================================================

🎯 Predicted Category: Literature & Fiction
📊 Confidence: 74.9%

🏆 Top 3 Predictions:
  1. Literature & Fiction                █████████████████████████████████████ 74.9%
  2. Children's Books                    █████ 10.1%
  3. Mystery, Thriller & Suspense        ██ 4.7%

📌 Test: Short Description

================================================================================
📚 Book: Short Description
================================================================================

🎯 Predicted Category: Literature & Fiction
📊 Confidence: 77.4%

🏆 Top 3 Predictions:
  1. Literature & Fiction                ██████████████████████████████████████ 77.4%
  2. Christian Books & Bibles            ███ 6.4%
  3. Mystery, Thriller & Suspense        ██ 4.9%

================================================================================
API SPEED BENCHMARK
================================================================================

Making 10 requests...
  Request 1: 2038.5ms
  Request 2: 2053.5ms
  Request 3: 2092.1ms
  Request 4: 2060.9ms
  Request 5: 2053.9ms
  Request 6: 2061.7ms
  Request 7: 2049.4ms
  Request 8: 2062.4ms
  Request 9: 2085.9ms
  Request 10: 2050.4ms

================================================================================
BENCHMARK RESULTS
================================================================================
Total requests:   10
Average time:     2060.9ms
Minimum time:     2038.5ms
Maximum time:     2092.1ms
Requests/second:  0.5

================================================================================
✓ ALL TESTS COMPLETED
================================================================================

================================================================================
TEST MENU
================================================================================
1. Test: Single book
2. Test: Multiple books
3. Test: Edge cases
4. Interactive mode
5. Speed benchmark
6. Run all tests
7. Exit

In [1]:
!python api.py 

^C
